packages/core/src/types.ts

```diff
@@ -1052,6 +1052,43 @@ export interface Typestate<TContext> {
   context: TContext;
 }
 
+type KeyOfUnion<T> = T extends unknown ? keyof T : never;
+
+type MergeUnion<T> = {
+  [K in keyof T]: T[K];
+} &
+  {
+    [K in Exclude<KeyOfUnion<T>, keyof T>]?: T extends Record<K, any>
+      ? T[K] | undefined
+      : never;
+  };
+
+export type CombineTypestates<
+  TStates extends Typestate<any>
+> = TStates['context'] extends infer C
+  ? TStates extends Typestate<any>
+    ? {
+        value: TStates['value'];
+        context: {
+          [K in keyof (MergeUnion<C> &
+            TStates['context'])]: K extends keyof TStates['context']
+            ? TStates['context'][K]
+            : undefined;
+        };
+      }
+    : never
+  : never;
+
+export type CombineTypestateTuples<
```

Comment:
when one would like to use this in the real app? `CombineTypestates` seems to be sufficient, especially given that it works on data shape that we document

Reply:
You're right :)

`CombineTypestateTuples` is the most terse way to define typestates, but I think `CombineTypestates` is more clear/obvious. I think this can be removed.